### PR TITLE
feat(frontend): major and minor

### PR DESCRIPTION
**Pull request by NathanJargon**
## Issues overlooked

- [x] #1 
     - Minor revision.
- [ ] #6 
     - added server.js, tried testing but has GET error
- [ ] #7 
     - Implemented csv parsing method (without sql) 
     - Implemented profiling and reporter class with example.
     - Implemented checkAvailabilityWithSchedule and getSubjectFrequencyAtDates methods.
- [ ] #8
     - Interface-to-JSdoc 
     - Problem: module specifier conflict with pdf2json (ignored to focus more on class)

## What has changed? [^1]

- parses compiled into class for reusable purposes 
- main.js minor revisions
- interface is a typescript (it is possible to just switch to JSDOC)
    - to make use of JSDOC (hover over the @param if you are using modern IDE for detailed documentation purposes)

## How to test the file? [^2]

```
git clone github.com/-------
install npm / install yarn
node main.js (please install node since package has no node version)
```

## What is the current output?

<details>

<summary>19/11/2023 (raw -- use prettify or something)</summary>
{
    "studentData": {
        "document_title": "CERTIFICATEOFREGISTRATION",
        "campus": "USTP CDO CAMPUS",
        "registration_no": "375433",
        "id": "2023302399",
        "name": "BONDOC, Nash Andrew Y",
        "gender": "M",
        "age": "18",
        "college": "COLLEGE OF INFORMATION TECHNOLOGY AND COMPUTING", 
        "department": "B.S. in Computer Science",
        "major": "N/A",
        "year_level": "1st Year - Baccalaureate",
        "curriculum": "BSCS-SY:2022-2023",
        "academic_year": "1st Semester AY 2023-2024",
        "scholarship": "N/A",
        "nationality": "Filipino",
        "contact": "N/A"
    },
    "subjectData": [
        {
            "subject": "Mathematics in the Modern World",
            "lecture": "3",
            "laboratory": "0",
            "credit": "3",
            "code": "MMW",
            "section": "CS1B",
            "schedule": [
                {
                    "instructor": "Christy  Jugan",
                    "weekday": "T",
                    "timeStart": "10:30 AM",
                    "timeEnd": "12:00 PM",
                    "room": ""
                },
                {
                    "instructor": "Christy  Jugan",
                    "weekday": "F",
                    "timeStart": "10:30 AM",
                    "timeEnd": "12:00 PM",
                    "room": ""
                }
            ]
        },
        {
            "subject": "Purposive Communication",
            "lecture": "3",
            "laboratory": "0",
            "credit": "3",
            "code": "PurCom",
            "section": "CS1B",
            "schedule": [
                {
                    "instructor": "Mary Lalainne  Park",
                    "weekday": "Th",
                    "timeStart": "6:00 PM",
                    "timeEnd": "7:30 PM",
                    "room": ""
                },
                {
                    "instructor": "Mary Lalainne  Park",
                    "weekday": "S",
                    "timeStart": "7:30 PM",
                    "timeEnd": "9:00 PM",
                    "room": ""
                }
            ]
        },
        {
            "subject": "The Contemporary World",
            "lecture": "3",
            "laboratory": "0",
            "credit": "3",
            "code": "TCW",
            "section": "CS1B",
            "schedule": [
                {
                    "instructor": "Ms. RC Tsini   Solis II",
                    "weekday": "W",
                    "timeStart": "6:00 PM",
                    "timeEnd": "9:00 PM",
                    "room": ""
                }
            ]
        },
        {
            "subject": "Readings in Philippine History",
            "lecture": "3",
            "laboratory": "0",
            "credit": "3",
            "code": "RPH",
            "section": "CS1B",
            "schedule": [
                {
                    "instructor": "June Merriam  . Peliño",
                    "weekday": "T",
                    "timeStart": "1:30 PM",
                    "timeEnd": "3:00 PM",
                    "room": ""
                },
                {
                    "instructor": "June Merriam  . Peliño",
                    "weekday": "F",
                    "timeStart": "1:30 PM",
                    "timeEnd": "3:00 PM",
                    "room": ""
                }
            ]
        },
        {
            "subject": "Movement Enhancement",
            "lecture": "2",
            "laboratory": "0",
            "credit": "2",
            "code": "PATH FIT 1",
            "section": "CS1B",
            "schedule": [
                {
                    "instructor": "Rhea Shainne A. Quiboy",
                    "weekday": "M",
                    "timeStart": "1:00 PM",
                    "timeEnd": "3:00 PM",
                    "room": ""
                }
            ]
        },
        {
            "subject": "Introduction to Computing",
            "lecture": "2",
            "laboratory": "1",
            "credit": "3",
            "code": "CS111",
            "section": "CS1B",
            "schedule": [
                {
                    "instructor": "Joie Ann  Mac",
                    "weekday": "Th",
                    "timeStart": "1:30 PM",
                    "timeEnd": "4:30 PM",
                    "room": ""
                },
                {
                    "instructor": "Joie Ann  Mac",
                    "weekday": "M",
                    "timeStart": "4:30 PM",
                    "timeEnd": "6:30 PM",
                    "room": ""
                }
            ]
        },
        {
            "subject": "Computer Programming 1",
            "lecture": "2",
            "laboratory": "1",
            "credit": "3",
            "code": "CS112",
            "section": "CS1B",
            "schedule": [
                {
                    "instructor": "Marylene S. Eder",
                    "weekday": "W",
                    "timeStart": "7:30 AM",
                    "timeEnd": "10:30 AM",
                    "room": "/09-306(ICT Lab6[45])"
                },
                {
                    "instructor": "Marylene S. Eder",
                    "weekday": "F",
                    "timeStart": "7:00 PM",
                    "timeEnd": "9:00 PM",
                    "room": ""
                }
            ]
        },
        {
            "subject": "Reserve Officers' Training Corps 1",
            "lecture": "3",
            "laboratory": "0",
            "credit": "3",
            "code": "NSTP101a",
            "section": "CS1B",
            "schedule": [
                {
                    "instructor": "Elgin T. Mansa",
                    "weekday": "S",
                    "timeStart": "8:00 AM",
                    "timeEnd": "12:00 PM",
                    "room": ""
                }
            ]
        }
    ],
    "frequencyMap": {
        "Mathematics in the Modern World": 2,
        "Purposive Communication": 1,
        "Readings in Philippine History": 2,
        "Movement Enhancement": 1,
        "Introduction to Computing": 1,
        "Computer Programming 1": 1,
        "Reserve Officers' Training Corps 1": 1
    },
    "availabilityData": {
        "S": {
            "availabilityCount": 2,
            "schedules": [
                {
                    "subject": "Purposive Communication",
                    "instructor": "Mary Lalainne  Park",
                    "timeStart": "7:30 PM",
                    "timeEnd": "9:00 PM",
                    "room": ""
                },
                {
                    "subject": "Reserve Officers' Training Corps 1",  
                    "instructor": "Elgin T. Mansa",
                    "timeStart": "8:00 AM",
                    "timeEnd": "12:00 PM",
                    "room": ""
                }
            ]
        },
        "M": {
            "availabilityCount": 2,
            "schedules": [
                {
                    "subject": "Movement Enhancement",
                    "instructor": "Rhea Shainne A. Quiboy",
                    "timeStart": "1:00 PM",
                    "timeEnd": "3:00 PM",
                    "room": ""
                },
                {
                    "subject": "Introduction to Computing",
                    "instructor": "Joie Ann  Mac",
                    "timeStart": "4:30 PM",
                    "timeEnd": "6:30 PM",
                    "room": ""
                }
            ]
        },
        "T": {
            "availabilityCount": 2,
            "schedules": [
                {
                    "subject": "Mathematics in the Modern World",     
                    "instructor": "Christy  Jugan",
                    "timeStart": "10:30 AM",
                    "timeEnd": "12:00 PM",
                    "room": ""
                },
                {
                    "subject": "Readings in Philippine History",      
                    "instructor": "June Merriam  . Peliño",
                    "timeStart": "1:30 PM",
                    "timeEnd": "3:00 PM",
                    "room": ""
                }
            ]
        },
        "F": {
            "availabilityCount": 3,
            "schedules": [
                {
                    "subject": "Mathematics in the Modern World",     
                    "instructor": "Christy  Jugan",
                    "timeStart": "10:30 AM",
                    "timeEnd": "12:00 PM",
                    "room": ""
                },
                {
                    "subject": "Readings in Philippine History",      
                    "instructor": "June Merriam  . Peliño",
                    "timeStart": "1:30 PM",
                    "timeEnd": "3:00 PM",
                    "room": ""
                },
                {
                    "subject": "Computer Programming 1",
                    "instructor": "Marylene S. Eder",
                    "timeStart": "7:00 PM",
                    "timeEnd": "9:00 PM",
                    "room": ""
                }
            ]
        }
    }
}
</details>

[^1]: Main.
[^2]: Secondary

> [!IMPORTANT]
> Always pull before pushing to avoid merge conflict! (And please create a fork before actually making a pull request for safety purposes) Please review changes, add comments in the code to clarify things. Thank you very much.

> [!CAUTION]
> fix pdf2json problem with its import or any client-side process will be hindered. `Failed to load resource: the server responded with a status of 404 (Not Found)` (after using "module")

### CODE REVIEW

**Comment by mihkuno** [comment]
- Use yarn
- readme.md hasn't been made yet mb
- Scheduler class is a client side function, it output frequency of subjects at certain dates
- Scheduler UI will be a static html file that displays such as a calender for further analysis
- Extractor class gets data from cor.pdf and google forms responses
- We'll stick with jsdocs, ty for the heads up

**Comment by NathanJargon** [comment]
> * Use yarn
> * readme.md hasn't been made yet mb
> * Scheduler class is a client side function, it output frequency of subjects at certain dates
> * Scheduler UI will be a static html file that displays such as a calender for further analysis
> * Extractor class gets data from cor.pdf and google forms responses
> * We'll stick with jsdocs, ty for the heads up

If the main dependency installer is yarn, then I'll remove `npm` in the next commit. Also, thanks for clarifying what needs to be tackled on the issues.

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 node_modules/
-misc/*.pdf
\ No newline at end of file
+misc/*.pdf
+index.html
```
Comment (mihkuno) [inline .gitignore:3]:
```suggestion
```

File: lib/extractor.js [hunk 2]
```diff
@@ -1,17 +1,13 @@
 import PDFParser from "pdf2json";
-
+import Scheduler from "./scheduler.js";
```
Comment (mihkuno) [inline lib/extractor.js:2]:
```suggestion
```
No interlinking of lib files

File: lib/extractor.js [hunk 3]
```diff
@@ -1,17 +1,13 @@
...
 // Public Functions
 
 async function getCorInfo(filePath) {
-
-    // TODO: check if document title is 'CERTIFICATEOFREGISTRATION'
     try {
         return await _getCorInfo(filePath);
-    } 
-    catch (error) {
-        // TODO: make a log file for all errors that occur
```
Comment (mihkuno) [inline lib/extractor.js:14]: create a file in log/log-*datetime*.log and append all that occurs here.

if it fails, 
- write a log as *datetime*:**FORMATERR** with an id to log/log.txt
- then put the file to log/*id.fileextension*

Comment (mihkuno) [reply lib/extractor.js:14]: a user might submit an edited cor.pdf file and the format is incorrect

File: lib/extractor.js [hunk 4]
```diff
@@ -1,17 +1,13 @@
 import PDFParser from "pdf2json";
-
+import Scheduler from "./scheduler.js";
 
 //
 // Public Functions
 
 async function getCorInfo(filePath) {
-
-    // TODO: check if document title is 'CERTIFICATEOFREGISTRATION'
```
Comment (mihkuno) [inline lib/extractor.js:9]: a user might submit a non pdf file without the cor contents
'CERTIFICATEOFREGISTRATION' should be checked

if it fails, 
- write a log as *datetime*:**NOTCOR** with an id to log/log.txt
- then put the file to log/*id.fileextension*

File: lib/extractor.js [hunk 5]
```diff
@@ -25,19 +21,46 @@ export { getCorInfo };
...
-        pdfParser.on("pdfParser_dataError", reject);
-        pdfParser.on("pdfParser_dataReady", (pdfData) => {
-            const data = _getPdfTextArray(pdfData);
-            const studentData = _getStudentData(data);
-            const subjectData = _getScheduleData(data);
-            resolve({ studentData, subjectData });
-        });
+        if (filePath.endsWith('.pdf')) {
+            _getPdfInfo(filePath, resolve, reject);
+        } else if (filePath.endsWith('.csv')) {
```
Comment (mihkuno) [inline lib/extractor.js:26]: there is no such cor with csv format

File: lib/extractor.js [hunk 6]
```diff
@@ -69,105 +92,9 @@ function _getPdfTextArray(pdfData) {
...
-                    room,
-                });
-            });
-
-            subjects.push(subject);
-        }
-    }
-
-    return subjects;
-}
```
Comment (mihkuno) [inline lib/extractor.js:173]: put it back

File: lib/profiler.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+class Profiler {
```
Comment (mihkuno) [inline lib/profiler.js:1]: - create a uml class diagram first to readme

File: lib/reporter.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+class Reporter {
```
Comment (mihkuno) [inline lib/reporter.js:1]: - create a uml class diagram first to readme

File: lib/scheduler.js [hunk 9]
```diff
@@ -0,0 +1,191 @@
+class Scheduler {
```
Comment (mihkuno) [inline lib/scheduler.js:1]: this is not what scheduler is for, it is a class that reads the schedule entity from the database.

queries their dates (sql prodecures to be added), and displays the frequency of students of each date.

- create a uml class diagram first to readme

File: lib/extractor.js [hunk 10]
```diff
@@ -25,19 +21,46 @@ export { getCorInfo };
...
+            _getPdfInfo(filePath, resolve, reject);
+        } else if (filePath.endsWith('.csv')) {
+            _getCsvInfo(filePath, resolve, reject);
+        } else {
+            reject(new Error('Unsupported file format'));
+        }
+    });
+}
+
+function _getPdfInfo(filePath, resolve, reject) {
```
Comment (mihkuno) [inline lib/extractor.js:34]: we're not reading anymore pdf's other than the cor

File: lib/extractor.js [hunk 11]
```diff
@@ -25,19 +21,46 @@ export { getCorInfo };
...
+        const data = _getPdfTextArray(pdfData);
+        const studentData = Scheduler.parseStudentData(data);
+        const subjectData = Scheduler.parseScheduleData(data);
+        const frequencyMap = Scheduler.getSubjectFrequencyAtDates(subjectData, datesToCheck);
+        const availabilityData = Scheduler.checkAvailabilityWithSchedule(subjectData, datesToCheck);
+        resolve({ studentData, subjectData, frequencyMap, availabilityData });
     });
 }
 
+/* CSV format is not supported yet
```
Comment (mihkuno) [inline lib/extractor.js:50]: the csv function is used to read the cached output responses of the google forms

File: lib/server.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
+import express from "express";
```
Comment (mihkuno) [inline lib/server.js:1]: @Illamapalooza is already working on this

File: main.js
Comment (mihkuno) [inline main.js:1]: all good here

**Review by mihkuno — COMMENTED** [review]
- Project structure met, some approved
- Some libs aren't what they are intended for, and shouldn't interlink to avoid complexity
- Make a uml class diagram to readme before implementing a class

Please make a seperate pr to each changes intended.

File: lib/scheduler.js (hunk 9 above)
Comment (NathanJargon) [reply lib/scheduler.js:1]: https://github.com/USTP-CS3/CRAFT-API/pull/12

**Pull request closed without merging**